haystack/preview/pipeline.py

```diff
@@ -53,33 +61,45 @@ def get_store(self, name: str) -> Store:
         :returns: the store
         """
         try:
-            return self.stores[name]
+            return self._stores[name]
         except KeyError as e:
             raise NoSuchStoreError(f"No store named '{name}' is connected to this pipeline.") from e
 
-    def run(self, data: Dict[str, ComponentInput], debug: bool = False) -> Dict[str, ComponentOutput]:
+    def add_component(self, name: str, instance: Any, store: Optional[str] = None) -> None:
         """
-        Wrapper on top of Canals Pipeline.run(). Adds the `stores` parameter to all nodes.
-
-        :params data: the inputs to give to the input components of the Pipeline.
-        :params parameters: a dictionary with all the parameters of all the components, namespaced by component.
-        :params debug: whether to collect and return debug information.
-        :returns A dictionary with the outputs of the output components of the Pipeline.
+        Make this component available to the pipeline. Components are not connected to anything by default:
+        use `Pipeline.connect()` to connect components together.
+
+        Component names must be unique, but component instances can be reused if needed.
+
+        If `store` has a value, the pipeline will also connect this component to the requested document store.
+        Note that only components that inherit from StoreAwareMixin can be connected to stores.
+
+        :param name: the name of the component.
+        :param instance: the component instance.
+        :param store: the store this component needs access to, if any.
+        :raises ValueError: if:
+            - a component with the same name already exists
+            - a component requiring a store didn't receive it
+            - a component that didn't expect a store received it
+        :raises PipelineValidationError: if the given instance is not a component
+        :raises NoSuchStoreError: if the given store name is not known to the pipeline
         """
-        # Get all nodes in this pipelines instance
-        for node_name in self.graph.nodes:
-            # Get node inputs
-            node = self.graph.nodes[node_name]["instance"]
-            input_params = find_input_sockets(node)
-
-            # If the node needs a store, adds the list of stores to its default inputs
-            if "stores" in input_params:
-                if not hasattr(node, "defaults"):
-                    setattr(node, "defaults", {})
-                node.defaults["stores"] = self.stores
-
-        # Run the pipeline
-        return super().run(data=data, debug=debug)
+        if isinstance(instance, StoreAwareMixin):
```

Comment:
What if this `instance` has already been added in this `Pipeline` with a different name? 
Shall we fail and force the user to pass the `store` again? What if there's some connection logic in the component?

Shall we maybe check if the `instance` already has a set `store`?

Reply:
I temporarily disabled this case by checking if store is already set. We'll consider this problem later